src/System.Management.Automation/engine/Subsystem/CommandPrediction/CommandPrediction.cs

```diff
@@ -0,0 +1,170 @@
+        }
+    }
+
+    /// <summary>
+    /// Provides a set of possible predictions for given input.
+    /// </summary>
+    public static class CommandPrediction
+    {
+        /// <summary>
+        /// Collect the predictive suggestions from registered predictors using the default timeout.
+        /// </summary>
+        /// <param name="ast">The <see cref="Ast"/> object from parsing the current command line input.</param>
+        /// <param name="astTokens">The <see cref="Token"/> objects from parsing the current command line input.</param>
+        /// <returns>A list of <see cref="PredictionResult"/> objects.</returns>
+        public static Task<List<PredictionResult>?> PredictInput(Ast ast, Token[] astTokens)
+        {
+            return PredictInput(ast, astTokens, millisecondsTimeout: 20);
+        }
+
+        /// <summary>
+        /// Collect the predictive suggestions from registered predictors using the specified timeout.
+        /// </summary>
+        /// <param name="ast">The <see cref="Ast"/> object from parsing the current command line input.</param>
+        /// <param name="astTokens">The <see cref="Token"/> objects from parsing the current command line input.</param>
+        /// <param name="millisecondsTimeout">The milliseconds to timeout.</param>
+        /// <returns>A list of <see cref="PredictionResult"/> objects.</returns>
+        public static async Task<List<PredictionResult>?> PredictInput(Ast ast, Token[] astTokens, int millisecondsTimeout)
+        {
+            Requires.Condition(millisecondsTimeout > 0, nameof(millisecondsTimeout));
+
+            var predictors = SubsystemManager.GetSubsystems<ICommandPredictor>();
+            if (predictors.Count == 0)
+            {
+                return null;
+            }
+
+            var context = new PredictionContext(ast, astTokens);
+            var tasks = new Task<PredictionResult?>[predictors.Count];
+            using var cancellationSource = new CancellationTokenSource();
+
+            for (int i = 0; i < predictors.Count; i++)
+            {
+                ICommandPredictor predictor = predictors[i];
+
+                tasks[i] = Task.Factory.StartNew(
+                    state => {
+                        var predictor = (ICommandPredictor)state!;
+                        List<PredictiveSuggestion>? texts = predictor.GetSuggestion(context, cancellationSource.Token);
+                        return texts?.Count > 0 ? new PredictionResult(predictor.Id, predictor.Name, texts) : null;
+                    },
+                    predictor,
+                    CancellationToken.None,
+                    TaskCreationOptions.DenyChildAttach,
+                    TaskScheduler.Default);
+            }
+
+            await Task.WhenAny(
+                Task.WhenAll(tasks),
+                Task.Delay(millisecondsTimeout, cancellationSource.Token)).ConfigureAwait(false);
+            cancellationSource.Cancel();
```

Comment:
> Why do we need the cancel for Task.Delay?

So that the timer is not triggered if it's not needed.

> And should we cancel predictor tasks after timeout?

The cancellation token is passed to `predictor.GetSuggestion`. It's up to the implementation to honor the cancellation.

Reply:
I wonder we are worried about Delay but not the implementations which can be very long unexpectedly and ignore the Cancel.

Reply:
An implementation can do whatever it wants. The important part is to make the public APIs exposed by `CommandPrediction` not affected by the potential wrong behaviors of the implementation, as much as possible. Here, we make sure a bad implementation cannot block `CommandPrediction.PredictInput` longer than `millisecondsTimeout`.

Of course, an implementation can still do bad things, such as exhausting all thread pool threads, deleting files on your computer :). But if you worry about that, then you should worry about all cmdlets out in the wild, because they all can do arbitrary crazy things in `BeginProcessing` and etc.

Reply:
I see you point. I want to understand can we use the cancellation token in Task.Create() above and avoid "exhausting all thread pool threads" or this can break anything in custom code?
If I think that the predictor uses a heavy/expensive service (AI/ML) I'd prefer to cancel extra tasks early.

Reply:
> I see you point. I want to understand can we use the cancellation token in Task.Create() above and avoid "exhausting all thread pool threads" or this can break anything in custom code?

That won't actually stop the code in the task from running, it'll just throw `TaskCanceledException` after it completes.  Any threads synchronously waiting for other threads to complete will still be holding thread pool slots.

Reply:
Passing in the same cancellation token to the task creation is good for 2 things:

1. The task can transition to the `Canceled` state properly. See the [task cancellation doc](https://docs.microsoft.com/en-us/dotnet/standard/parallel-programming/task-cancellation) for details, and quote part of the content here:
   > When a task instance observes an OperationCanceledException thrown by user code, it compares the exception's token to its associated token (the one that was passed to the API that created the Task). If they are the same and the token's IsCancellationRequested property returns true, the task interprets this as acknowledging cancellation and transitions to the Canceled state.

2. It helps to proactively remove the task from the scheduler's queue if it's not yet started. See the [code here](https://source.dot.net/#System.Private.CoreLib/Task.cs,594)

When the task is already started, then [cancellation is cooperative and is not forced on the listener](https://docs.microsoft.com/en-us/dotnet/standard/threading/cancellation-in-managed-threads).

Reply:
I guess the expectation here is that predictor implementations return not complete result list but results by one and async. I am not sure whether this is acceptable or no.